filebeat/input/kafka/input.go

```diff
@@ -0,0 +1,293 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package kafka
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/Shopify/sarama"
+
+	"github.com/elastic/beats/filebeat/channel"
+	"github.com/elastic/beats/filebeat/input"
+	"github.com/elastic/beats/libbeat/beat"
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/common/kafka"
+	"github.com/elastic/beats/libbeat/logp"
+
+	"github.com/pkg/errors"
+)
+
+func init() {
+	err := input.Register("kafka", NewInput)
+	if err != nil {
+		panic(err)
+	}
+}
+
+// Input contains the input and its config
+type kafkaInput struct {
+	config          kafkaInputConfig
+	saramaConfig    *sarama.Config
+	context         input.Context
+	outlet          channel.Outleter
+	saramaWaitGroup sync.WaitGroup // indicates a sarama consumer group is active
+	log             *logp.Logger
+	runOnce         sync.Once
+}
+
+// NewInput creates a new kafka input
+func NewInput(
+	cfg *common.Config,
+	connector channel.Connector,
+	inputContext input.Context,
+) (input.Input, error) {
+
+	out, err := connector.ConnectWith(cfg, beat.ClientConfig{
```

Comment:
Added the CloseRef parameter (which also let me remove the extra Once for shutdown). I'm not sure how to use `WaitClose` effectively though, since I don't see any code that uses it currently. Are you suggesting I make it configurable as part of the input via `kafkaInputConfig`?

Reply:
for WaitClose you only have to configure a duration. When the outlet is closed, it blocks after all events have been ACKed, or until the configured duration passed. During this period ACKs are still received, but no events can be published anymore.

> Are you suggesting I make it configurable as part of the input via kafkaInputConfig?

Yes.